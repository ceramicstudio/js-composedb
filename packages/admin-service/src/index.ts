export { router } from './router.js'
export type { Router } from './router.js'
export { GraphQLQueryCodec, Service } from './service.js'
export type { GraphQLResult, ServiceConfig, ServiceParams } from './service.js'
