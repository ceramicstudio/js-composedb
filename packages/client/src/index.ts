/**
 * High-level ComposeDB client, based on the {@linkcode runtime ComposeDB runtime}.
 *
 * ## Installation
 *
 * ```sh
 * npm install @composedb/client
 * ```
 *
 * @module client
 */

export { ComposeClient, type ComposeClientParams } from './client.js'
