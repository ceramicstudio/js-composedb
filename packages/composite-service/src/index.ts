export { router } from './router.js'
export type { CreateModelInput, LoadModelInput, SaveCompositeInput, Router } from './router.js'
export { Service } from './service.js'
export type { GraphQLResult, SavedComposite, ServiceParams } from './service.js'
