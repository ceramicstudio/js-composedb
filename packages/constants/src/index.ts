export const NODE_ID_FIELD = 'id'

export const VIEWER_ID_HEADER = 'ComposeDB-ViewerID'
