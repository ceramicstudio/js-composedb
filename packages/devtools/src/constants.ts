export const DOC_ID_FIELD = 'id'
