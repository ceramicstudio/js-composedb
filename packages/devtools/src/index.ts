/**
 * Development tools library.
 *
 * ## Installation
 *
 * ```sh
 * npm install --dev @composedb/devtools
 * ```
 *
 * @module devtools
 */

export * from './composite.js'
export * from './formats/json.js'
export { createRuntimeDefinition, getName } from './formats/runtime.js'
export { createAbstractCompositeDefinition } from './schema/compiler.js'
export * from './schema/types.js'
export * from './test-utils.js'
export * from './types.js'
export * from './utils.js'
