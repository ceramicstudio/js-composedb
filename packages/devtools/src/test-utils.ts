import type { ModelDefinition, ModelViewsDefinition } from '@ceramicnetwork/stream-model'
import type { InternalCompositeDefinition } from '@composedb/types'

import { createAbstractCompositeDefinition } from './schema/compiler.js'
import type { AbstractCompositeDefinition } from './schema/types.js'

/** @internal */
export function mockDefinition(
  definition: AbstractCompositeDefinition,
  providedModels: Record<string, ModelDefinition> = {}
): InternalCompositeDefinition {
  const models: Record<string, ModelDefinition> = {}
  const modelsViews: Record<string, ModelViewsDefinition> = {}

  for (const abstractModel of Object.values(definition.models)) {
    if (abstractModel.action === 'create') {
      const definition = abstractModel.model
      models[`${definition.name}ID`] = definition
    } else {
      const definition = providedModels[abstractModel.id]
      if (definition == null) {
        throw new Error(`Missing provided model ${abstractModel.id}`)
      }
      models[abstractModel.id] = definition
      modelsViews[abstractModel.id] = abstractModel.views
    }
  }

  return {
    version: '1.0',
    commonEmbeds: definition.commonEmbeds,
    models,
    views: { models: modelsViews },
  }
}

/** @internal */
export function mockDefinitionFromSchema(
  schema: string,
  providedModels?: Record<string, ModelDefinition>
): InternalCompositeDefinition {
  return mockDefinition(createAbstractCompositeDefinition(schema), providedModels)
}
