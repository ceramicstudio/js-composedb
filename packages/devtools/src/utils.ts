import type { ModelRelationViewDefinition, ModelViewDefinition } from '@ceramicnetwork/stream-model'
import type { StreamRef } from '@ceramicnetwork/streamid'
import type { RuntimeRelation, RuntimeRelationSource, RuntimeViewField } from '@composedb/types'

/** @internal */
export function streamIDToString(id: StreamRef | string): string {
  return typeof id === 'string' ? id : id.toString()
}

/** @internal */
export function applyMap<
  M extends Record<string, unknown>,
  V extends M[keyof M] = M[keyof M],
  R = unknown
>(inputs: M, callFunc: (input: V) => R): Record<keyof M, R> {
  return Object.entries(inputs).reduce((acc, [key, value]) => {
    acc[key as keyof M] = callFunc(value as V)
    return acc
  }, {} as Record<keyof M, R>)
}

/** @internal */
export async function promiseMap<
  M extends Record<string, unknown>,
  V extends M[keyof M] = M[keyof M],
  R = unknown
>(inputs: M, callFunc: (input: V) => Promise<R>): Promise<Record<keyof M, R>> {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
  const results = await Promise.all(Object.values(inputs).map((value) => callFunc(value as any)))
  return Object.keys(inputs).reduce((acc, key, i) => {
    acc[key as keyof M] = results[i]
    return acc
  }, {} as Record<keyof M, R>)
}

type RelationViewType = ModelRelationViewDefinition['type']

const RELATION_VIEW_SOURCES: Record<RelationViewType, RuntimeRelationSource> = {
  relationDocument: 'document',
  relationFrom: 'queryConnection',
  relationCountFrom: 'queryCount',
}

const RELATION_VIEW_TYPES = Object.keys(RELATION_VIEW_SOURCES)

/** @internal */
export function assertRelationViewDefinition(
  view: ModelViewDefinition
): ModelRelationViewDefinition {
  if (!RELATION_VIEW_TYPES.includes(view.type)) {
    throw new Error(`Invalid relation view: ${view.type}`)
  }
  return view as ModelRelationViewDefinition
}

/** @internal */
export function viewDefinitionToRuntimeRelation(view: ModelViewDefinition): RuntimeRelation {
  const relation = assertRelationViewDefinition(view)
  return {
    source: RELATION_VIEW_SOURCES[relation.type],
    model: relation.model,
    property: relation.property,
  }
}

/** @internal */
export function viewDefinitionToRuntime(view: ModelViewDefinition): RuntimeViewField {
  switch (view.type) {
    case 'documentAccount':
    case 'documentVersion':
      return { type: 'view', viewType: view.type }
    case 'relationCountFrom':
    case 'relationDocument':
    case 'relationFrom':
      return {
        type: 'view',
        viewType: 'relation',
        relation: viewDefinitionToRuntimeRelation(view),
      }
    default:
      // @ts-ignore unexpected view type
      throw new Error(`Unsupported view type: ${view.type as string}`)
  }
}

/** @internal */
export function viewRuntimeToModel(field: RuntimeViewField): ModelViewDefinition {
  if (field.viewType === 'relation') {
    const { model, property, source } = field.relation
    switch (source) {
      case 'document':
        return { type: 'relationDocument', model, property }
      case 'queryConnection':
        return { type: 'relationFrom', model, property }
      case 'queryCount':
        return { type: 'relationCountFrom', model, property }
    }
  }
  return { type: field.viewType }
}
