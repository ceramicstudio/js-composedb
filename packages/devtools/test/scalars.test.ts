import { extraScalarsSchema } from '@composedb/test-schemas'

import { createRuntimeDefinition, mockDefinitionFromSchema } from '../src'

describe('scalars', () => {
  test('support extra scalars', () => {
    const scalarsDefinition = mockDefinitionFromSchema(extraScalarsSchema)
    expect(scalarsDefinition).toMatchSnapshot()
    const runtime = createRuntimeDefinition(scalarsDefinition)
    expect(runtime).toMatchSnapshot()
  })
})
