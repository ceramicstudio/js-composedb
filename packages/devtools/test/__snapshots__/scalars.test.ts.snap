// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scalars support extra scalars 1`] = `
{
  "commonEmbeds": [],
  "models": {
    "ExtraScalarsID": {
      "accountRelation": {
        "type": "list",
      },
      "description": "Kitchen sink of added scalar types",
      "name": "ExtraScalars",
      "relations": {},
      "schema": {
        "$defs": {
          "CeramicCommitID": {
            "maxLength": 200,
            "title": "CeramicCommitID",
            "type": "string",
          },
          "CeramicStreamID": {
            "maxLength": 100,
            "title": "CeramicStreamID",
            "type": "string",
          },
          "ChainAgnosticAccountID": {
            "maxLength": 106,
            "title": "ChainAgnosticAccountID",
            "type": "string",
          },
          "ChainAgnosticChainID": {
            "maxLength": 41,
            "title": "ChainAgnosticChainID",
            "type": "string",
          },
          "GraphQLCountryCode": {
            "maxLength": 2,
            "pattern": "^(AD|AE|AF|AG|AI|AL|AM|AO|AQ|AR|AS|AT|AU|AW|AX|AZ|BA|BB|BD|BE|BF|BG|BH|BI|BJ|BL|BM|BN|BO|BQ|BR|BS|BT|BV|BW|BY|BZ|CA|CC|CD|CF|CG|CH|CI|CK|CL|CM|CN|CO|CR|CU|CV|CW|CX|CY|CZ|DE|DJ|DK|DM|DO|DZ|EC|EE|EG|EH|ER|ES|ET|FI|FJ|FK|FM|FO|FR|GA|GB|GD|GE|GF|GG|GH|GI|GL|GM|GN|GP|GQ|GR|GS|GT|GU|GW|GY|HK|HM|HN|HR|HT|HU|ID|IE|IL|IM|IN|IO|IQ|IR|IS|IT|JE|JM|JO|JP|KE|KG|KH|KI|KM|KN|KP|KR|KW|KY|KZ|LA|LB|LC|LI|LK|LR|LS|LT|LU|LV|LY|MA|MC|MD|ME|MF|MG|MH|MK|ML|MM|MN|MO|MP|MQ|MR|MS|MT|MU|MV|MW|MX|MY|MZ|NA|NC|NE|NF|NG|NI|NL|NO|NP|NR|NU|NZ|OM|PA|PE|PF|PG|PH|PK|PL|PM|PN|PR|PS|PT|PW|PY|QA|RE|RO|RS|RU|RW|SA|SB|SC|SD|SE|SG|SH|SI|SJ|SK|SL|SM|SN|SO|SR|SS|ST|SV|SX|SY|SZ|TC|TD|TF|TG|TH|TJ|TK|TL|TM|TN|TO|TR|TT|TV|TW|TZ|UA|UG|UM|US|UY|UZ|VA|VC|VE|VG|VI|VN|VU|WF|WS|YE|YT|ZA|ZM|ZW)$",
            "title": "GraphQLCountryCode",
            "type": "string",
          },
          "GraphQLDID": {
            "maxLength": 100,
            "pattern": "^did:[a-zA-Z0-9.!#$%&'*+\\/=?^_\`{|}~-]+:[a-zA-Z0-9.!#$%&'*+\\/=?^_\`{|}~-]+$",
            "title": "GraphQLDID",
            "type": "string",
          },
          "GraphQLDate": {
            "format": "date",
            "maxLength": 100,
            "title": "GraphQLDate",
            "type": "string",
          },
          "GraphQLDateTime": {
            "format": "date-time",
            "maxLength": 100,
            "title": "GraphQLDateTime",
            "type": "string",
          },
          "GraphQLLatitude": {
            "maximum": 90,
            "minimum": -90,
            "title": "GraphQLLatitude",
            "type": "number",
          },
          "GraphQLLocalDate": {
            "maxLength": 100,
            "pattern": "^\\d{4}-\\d{2}-\\d{2}$",
            "title": "GraphQLLocalDate",
            "type": "string",
          },
          "GraphQLLocalTime": {
            "maxLength": 100,
            "pattern": "^([0-1][0-9]|2[0-3]):([0-5][0-9])(:[0-5][0-9](\\.\\d{3})?)?$",
            "title": "GraphQLLocalTime",
            "type": "string",
          },
          "GraphQLLocale": {
            "maxLength": 100,
            "pattern": "^(((en-GB-oed|i-ami|i-bnn|i-default|i-enochian|i-hak|i-klingon|i-lux|i-mingo|i-navajo|i-pwn|i-tao|i-tay|i-tsu|sgn-BE-FR|sgn-BE-NL|sgn-CH-DE)|(art-lojban|cel-gaulish|no-bok|no-nyn|zh-guoyu|zh-hakka|zh-min|zh-min-nan|zh-xiang))|((([A-Za-z]{2,3}(-([A-Za-z]{3}(-[A-Za-z]{3}){0,2}))?)|[A-Za-z]{4}|[A-Za-z]{5,8})(-([A-Za-z]{4}))?(-([A-Za-z]{2}|[0-9]{3}))?(-([A-Za-z0-9]{5,8}|[0-9][A-Za-z0-9]{3}))*(-([0-9A-WY-Za-wy-z](-[A-Za-z0-9]{2,8})+))*(-(x(-[A-Za-z0-9]{1,8})+))?)|(x(-[A-Za-z0-9]{1,8})+))$",
            "title": "GraphQLLocale",
            "type": "string",
          },
          "GraphQLLongitude": {
            "maximum": 180,
            "minimum": -180,
            "title": "GraphQLLongitude",
            "type": "number",
          },
          "GraphQLTime": {
            "format": "time",
            "maxLength": 100,
            "title": "GraphQLTime",
            "type": "string",
          },
          "GraphQLTimeZone": {
            "maxLength": 100,
            "pattern": "^(?:[A-Za-z0-9_]|(?:%[0-9A-Fa-f]{2}))+",
            "title": "GraphQLTimeZone",
            "type": "string",
          },
          "GraphQLURI": {
            "format": "uri",
            "maxLength": 100,
            "title": "GraphQLURI",
            "type": "string",
          },
          "GraphQLUtcOffset": {
            "maxLength": 100,
            "pattern": "^([+-]?)(\\d{2}):(\\d{2})$",
            "title": "GraphQLUtcOffset",
            "type": "string",
          },
          "InterPlanetaryCID": {
            "maxLength": 100,
            "title": "InterPlanetaryCID",
            "type": "string",
          },
        },
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "additionalProperties": false,
        "properties": {
          "accountID": {
            "$ref": "#/$defs/ChainAgnosticAccountID",
          },
          "chainID": {
            "$ref": "#/$defs/ChainAgnosticChainID",
          },
          "cid": {
            "$ref": "#/$defs/InterPlanetaryCID",
          },
          "commitID": {
            "$ref": "#/$defs/CeramicCommitID",
          },
          "countryCode": {
            "$ref": "#/$defs/GraphQLCountryCode",
          },
          "date": {
            "$ref": "#/$defs/GraphQLDate",
          },
          "dateTime": {
            "$ref": "#/$defs/GraphQLDateTime",
          },
          "did": {
            "$ref": "#/$defs/GraphQLDID",
          },
          "latitude": {
            "$ref": "#/$defs/GraphQLLatitude",
          },
          "localDate": {
            "$ref": "#/$defs/GraphQLLocalDate",
          },
          "localTime": {
            "$ref": "#/$defs/GraphQLLocalTime",
          },
          "locale": {
            "$ref": "#/$defs/GraphQLLocale",
          },
          "longitude": {
            "$ref": "#/$defs/GraphQLLongitude",
          },
          "streamID": {
            "$ref": "#/$defs/CeramicStreamID",
          },
          "time": {
            "$ref": "#/$defs/GraphQLTime",
          },
          "timeZone": {
            "$ref": "#/$defs/GraphQLTimeZone",
          },
          "uri": {
            "$ref": "#/$defs/GraphQLURI",
          },
          "utcOffset": {
            "$ref": "#/$defs/GraphQLUtcOffset",
          },
        },
        "type": "object",
      },
      "version": "1.0",
      "views": {},
    },
  },
  "version": "1.0",
  "views": {
    "models": {},
  },
}
`;

exports[`scalars support extra scalars 2`] = `
{
  "accountData": {
    "extraScalarsList": {
      "name": "ExtraScalars",
      "type": "connection",
    },
  },
  "enums": {},
  "models": {
    "ExtraScalars": {
      "accountRelation": {
        "type": "list",
      },
      "id": "ExtraScalarsID",
    },
  },
  "objects": {
    "ExtraScalars": {
      "accountID": {
        "required": false,
        "type": "accountid",
      },
      "chainID": {
        "required": false,
        "type": "chainid",
      },
      "cid": {
        "required": false,
        "type": "cid",
      },
      "commitID": {
        "required": false,
        "type": "commitid",
      },
      "countryCode": {
        "required": false,
        "type": "countrycode",
      },
      "date": {
        "required": false,
        "type": "date",
      },
      "dateTime": {
        "required": false,
        "type": "datetime",
      },
      "did": {
        "required": false,
        "type": "did",
      },
      "latitude": {
        "required": false,
        "type": "float",
      },
      "localDate": {
        "required": false,
        "type": "localdate",
      },
      "localTime": {
        "required": false,
        "type": "localtime",
      },
      "locale": {
        "required": false,
        "type": "locale",
      },
      "longitude": {
        "required": false,
        "type": "float",
      },
      "streamID": {
        "required": false,
        "type": "streamid",
      },
      "time": {
        "required": false,
        "type": "time",
      },
      "timeZone": {
        "required": false,
        "type": "timezone",
      },
      "uri": {
        "required": false,
        "type": "uri",
      },
      "utcOffset": {
        "required": false,
        "type": "utcoffset",
      },
    },
  },
}
`;
